File: tools/extraction/binance/src/processor_registry.rs
use std::error::Error;
use async_trait::async_trait;

// Define a trait for document processors with proper async support
#[async_trait]
pub trait DocumentProcessor {
    async fn process_docs(&self) -> Result<(u32, String, String), Box<dyn Error + Send + Sync>>;
    fn get_output_filename(&self) -> String;
}

// Placeholder processor implementation - make public
pub struct ExchangeProcessor {
    exchange_id: String,
}

impl ExchangeProcessor {
    pub fn new(exchange_id: &str) -> Self {
        Self {
            exchange_id: exchange_id.to_string(),
        }
    }
}

#[async_trait]
impl DocumentProcessor for ExchangeProcessor {
    async fn process_docs(&self) -> Result<(u32, String, String), Box<dyn Error + Send + Sync>> {
        // Placeholder implementation
        // Returns (token_count, timestamp, market_name)
        Ok((1000, "2023-11-01".to_string(), self.exchange_id.clone()))
    }

    fn get_output_filename(&self) -> String {
        format!("{}_documentation.md", self.exchange_id)
    }
}

// Factory function to create processors for a specific exchange
pub fn create_processors_by_exchange(exchange_type: &str) -> Vec<Box<dyn DocumentProcessor>> {
    match exchange_type {
        "binanceusdm" => {
            // Return processors specific to Binance USDM
            vec![Box::new(ExchangeProcessor::new(exchange_type))]
        },
        "binancespot" => {
            // Return processors specific to Binance Spot
            vec![Box::new(ExchangeProcessor::new(exchange_type))]
        },
        "binancecoinm" => {
            // Return processors specific to Binance CoinM
            vec![Box::new(ExchangeProcessor::new(exchange_type))]
        },
        "binanceoptions" => {
            // Return processors specific to Binance Options
            vec![Box::new(ExchangeProcessor::new(exchange_type))]
        },
        _ => {
            // Return empty vector for unknown exchange type
            vec![]
        }
    }
}

File: venues/binancespot/src/processorRegistry.js
import { BaseProcessor } from './baseProcessor.js'
import {
  BinanceSpotFIXProcessor,
  BinanceSpotPrivateRESTProcessor,
  BinanceSpotPrivateWebSocketProcessor,
  BinanceSpotPublicRESTProcessor,
  BinanceSpotPublicWebSocketProcessor,
  BinanceSpotSBEProcessor,
} from './binanceSpot.js'

/**
 * Registry for creating and managing document processors
 */
export class ProcessorRegistry {
  /**
   * Creates processors based on the exchange type
   * @param {string} exchangeType - The type of exchange to create processors for
   * @returns {BaseProcessor[]} Array of processors for the specified exchange
   * @throws {Error} If the exchange type is unknown
   */
  static createProcessorsByExchange(exchangeType) {
    switch (exchangeType) {
      case 'binancespot':
        return [
          new BinanceSpotPublicRESTProcessor(),
          new BinanceSpotPrivateRESTProcessor(),
          new BinanceSpotPublicWebSocketProcessor(),
          new BinanceSpotPrivateWebSocketProcessor(),
          new BinanceSpotFIXProcessor(),
          new BinanceSpotSBEProcessor(),
        ]
      default:
        throw new Error(`Unknown exchange type: ${exchangeType}`)
    }
  }
}
